core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -1054,6 +1054,23 @@ sealed trait ZIO[-R, +E, +A] extends Serializable { self =>
     schedule.initial.flatMap(loop(None, _))
   }
 
+  /**
+   * Repeats this effect until the partial function succeeds on the result.
+   */
+  final def repeatUntil[B](f: PartialFunction[A, B]): ZIO[R, E, B] =
+    self.flatMap { a =>
+      f.lift(a) match {
+        case Some(value) => ZIO.succeed(value)
+        case None        => repeatUntil(f)
+      }
+    }
+
+  /**
+   * Effectful version of `repeatUntil`. Will repeat until `f` succeeds
+   */
+  final def repeatUntilM[R1 <: R, E2, B](f: A => ZIO[R1, E2, B]): ZIO[R1, E2, B] =
```

Comment:
Should the result be typed `ZIO[R1, Nothing, B]`? The function will never fail.

Also, this is essentially `flatMap(f).eventually`. Does this justify a dedicated combinator?

Reply:
It should. The main use case was using the console to read input and make sure it's valid:

```scala
for {
    _ <- putStrLn("Please enter a time in 24 hour format (hh:mm): ")
    userTime <- getStrLn.repeatUntil(in => Task.effect(in.toString).tapError(_ => putStrLn("Please enter a valid value"))
  } yield userTime
```

Does this justify a dedicated combinator? I'm not sure :)

Reply:
Ok, let's wait for another opinion :-)

Reply:
I talked with @jdegoes about this during the workshop, perhaps he can add his 2 cents.

Reply:
The return type should definitely omit the final error, and we can implement it using `(self >>= f).eventually`. 

The symmetry with `repeatUntil` is nice, although perhaps we could make it closer still by allowing `PartialFunction[A, ZIO[R1, E2, B]]`. The implementation of this one is also more complex and possibly justifying its existence.

Thoughts?

Reply:
SGTM. I've reworked the implementation, WDYT now? @iravid @jdegoes @neko-kai

Reply:
Did you mean to convert `repeatUntilM` to take a `PartialFunction`?

Reply:
I.e. `repeatUntil` just uses a plain partial function, `repeatUntilM` allows evaluating an effectful partial function

Reply:
@iravid If I understood correctly, I think he meant for `repeatUntil` to take a `PartialFunction[A, ZIO[R1, E1, B]]` and for `repeatUntilM` to take a total function `A => ZIO[R1, E1, B]`